src/sqlfluff/core/config.py

```diff
@@ -503,13 +504,32 @@ def __init__(
         # NB: We import here to avoid a circular references.
         from sqlfluff.core.dialects import dialect_selector
 
-        self._configs["core"]["dialect_obj"] = dialect_selector(
-            self._configs["core"]["dialect"]
-        )
+        dialect: Optional[str] = self._configs["core"]["dialect"]
+        if dialect is not None:
+            self._configs["core"]["dialect_obj"] = dialect_selector(
+                self._configs["core"]["dialect"]
+            )
+        elif require_dialect:
+            self.verify_dialect_specified()
```

Comment:
This reads a little weird

```
if dialect is selected
   do something
else: # (so dialect is obviously NOT selected)
  verify if dialect is selected # even though we know it is not!!!
```

Would it be better to just drop the `elif`? Would also allow us to check if the above `dialect_selector` call failed.

```py
        if dialect is not None:
            self._configs["core"]["dialect_obj"] = dialect_selector(
                self._configs["core"]["dialect"]
            )
        self.verify_dialect_specified()
```

Reply:
I'm worried this is one of those situations where I'm going to try and clean up the code but break it instead. Let's leave this as is, okay? Maybe another time...

Reply:
We need to check the `require_dialect` flag, because if it's false, we don't want to call `verify_dialect_specified()` because it would definitely fail, which we don't want in this case.

There may be cleaner ways to do parts of this at the code level. Trying to retrofit smarter behavior into code that was built with certain assumptions is often messy on the first go. I think it works, and more worried about breaking it than leaving it a bit ugly.

Reply:
Grand. Not a big deal.